src/reconcile/rec_child.c

```diff
@@ -17,86 +17,53 @@ __rec_child_deleted(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *ref, WT_C
 {
     WT_PAGE_DELETED *page_del;
 
+    *statep = WT_CHILD_IGNORE;
+
+    /*
+     * The complicated case is a fast-delete which may not be visible or stable. Otherwise, discard
+     * any underlying disk blocks and don't write anything.
+     */
     page_del = ref->ft_info.del;
+    if (page_del == NULL)
+        return (ref->addr == NULL ? 0 : __wt_ref_block_free(session, ref));
 
     /*
-     * Internal pages with child leaf pages in the WT_REF_DELETED state are a special case during
-     * reconciliation. First, if the deletion was a result of a session truncate call, the deletion
-     * may not be visible to us. In that case, we proceed as with any change not visible during
-     * reconciliation by ignoring the change for the purposes of writing the internal page.
-     *
-     * In this case, there must be an associated page-deleted structure, and it holds the
-     * transaction ID we care about.
+     * The fast-delete may not yet be visible to us. In that case, we proceed as with any change not
+     * visible during reconciliation by ignoring the change for the purposes of writing the internal
+     * page.
      *
-     * In some cases, there had better not be any updates we can't see.
-     *
-     * A visible update to be in READY state (i.e. not in LOCKED or PREPARED state), for truly
-     * visible to others.
+     * We expect the page to be clean after reconciliation. If there are invisible updates, abort
+     * eviction.
      */
-    if (F_ISSET(r, WT_REC_CLEAN_AFTER_REC | WT_REC_VISIBILITY_ERR) && page_del != NULL &&
-      __wt_page_del_active(session, ref, false)) {
+    if (__wt_page_del_active(session, ref, false)) {
         if (F_ISSET(r, WT_REC_VISIBILITY_ERR))
             WT_RET_PANIC(session, EINVAL, "reconciliation illegally skipped an update");
-        return (__wt_set_return(session, EBUSY));
+        if (F_ISSET(r, WT_REC_CLEAN_AFTER_REC))
+            return (__wt_set_return(session, EBUSY));
+        *statep = WT_CHILD_ORIGINAL;
+        return (0);
```

Comment:
@quchenhao, @kommiharibabu, my second question is this change. The code I have in the branch now isn't quite right, it occasionally fails in `test_truncate03.py`, when under load. The failure stack is
```
stack is eviction ->
    reconcile ->
    wt_rec_row_int ->
    wt_rec_child_modify ->
    __wt_page_del_active(session, ref, false) ->
    wt_page_del_active ->
    wt_txn_visible ->
    txn_visible_id:

WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) ||
    session->dhandle->checkpoint != NULL);
```

I think the previous version's flag checks:
```
F_ISSET(r, WT_REC_CLEAN_AFTER_REC | WT_REC_VISIBILITY_ERR)
```
prevented this core dump, but I don't off-hand see how that would be the case, and I'm not sure what I should be doing here if eviction calls this code with (or without), acquiring a snapshot.

Can either of you explain this for me?

Thanks!1

Reply:
If eviction doesn't have a snapshot, we need to compare the txnid with r->last_running like how we do it in __wt_rec_upd_select. I'm not sure how the flags prevented it from crashing before.

Reply:
I don't think the above problem relates to this PR. @AlexeyAnisimovAU reproduced the similar problem on the develop branch. More details are available in WT-8379.

Reply:
Thanks, @kommiharibabu. There's now a ticket for this one, WT-8604, and I agree, it's not a problem here.